src/backend/crossterm.rs

```diff
@@ -127,6 +119,194 @@ where
     }
 }
 
+impl CrosstermBackend<std::io::Stdout> {
+    /// Creates a new `CrosstermBackend` with `std::io::stdout`.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::prelude::*;
+    ///
+    /// let backend = CrosstermBackend::stdout();
+    /// ```
+    pub fn stdout() -> Self {
+        Self::new(std::io::stdout())
+    }
+}
+
+impl CrosstermBackend<std::io::Stderr> {
+    /// Creates a new `CrosstermBackend` with `std::io::stderr`.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::prelude::*;
+    ///
+    /// let backend = CrosstermBackend::stderr();
+    /// ```
+    pub fn stderr() -> Self {
+        Self::new(std::io::stderr())
+    }
+}
+
+impl<W: Write> CrosstermBackend<W> {
+    /// Converts the `CrosstermBackend` into a [`Terminal`] instance.
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// use ratatui::backend::CrosstermBackend;
+    ///
+    /// let terminal = CrosstermBackend::stdout()
+    ///     .with_raw_mode()?
+    ///     .with_alternate_screen()?
+    ///     .with_mouse_support()?
+    ///     .into_terminal()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    pub fn into_terminal(self) -> io::Result<Terminal<Self>> {
```

Comment:
I dont think I like this approach as it creates two ways of creating a terminal.
Either `Terminal::new(backend)` or `backend.into_terminal()`. The second one needs to be implemented again and again.

I think there should be exactly one way to do something, not multiple ones.

Reply:
I'm fine with this. Its purpose is clear and obvious. It makes it easy for the common use case to be a single one liner. Reading through a bunch of examples of into_ vs to_, I think this actually makes more sense as a `to_` method as it's not strictly a conversion. There are many many instances of being able to do the same thing multiple ways in the std lib.

One change I'd make that I think works nicer though is to move the _with_defaults part out of creating a terminal. 
```rust
// So instead of
let terminal = Crossterm::stdout().into_terminal_with_defaults();
// change to
let terminal = Crossterm::stdout().with_default_config().into_terminal();
```

Reply:
Also moved `to_terminal` to Backend trait.

Reply:
`into` moves, `to` does not move. This moves → into.

Still don't like different ways to do the same.

Reply:
https://rust-lang.github.io/api-guidelines/naming.html#c-conv

> Conversions prefixed as_ and into_ typically decrease abstraction, either exposing a view into the underlying representation (as) or deconstructing data into its underlying representation (into). Conversions prefixed to_, on the other hand, typically stay at the same level of abstraction but do some work to change from one representation to another.

This feels more like a to_ method to my understanding as it doesn't convert, it just wraps.

Reply:
While attempting to convert all the examples to use this, often github CoPilot would attempt to complete `let terminal = CrosstermBackend::stdout_with_defaults()?` with `.into_terminal()?;`. This sort of implies that there's enough statistical evidence that into is a better name (like AI or hate it, often LLM hallucinations tell us what the obvious thing should be even if it's not that thing).

Upshot - I was wrong here - and will change it to into_terminal.

Reply:
408e37ea6

Reply:
Given your cited paragraph, `as_terminal` might be another good naming. `into` also differentiates from `to` by moving into or out of something else. `to` works with reference while `into` moves the ownership.

---

Conversion methods can be fine and great, but ratatui seems to use them way too much in my taste. The Into / From traits are easy foot guns in refactorings, especially when they might use expensive logic. But here I am not against the conversion methods, I am against the multiple ways of doing things.
Currently, there are a lot of test coverage warnings because there are multiple ways of doing the same thing. Tests use one but not the other way. These can easily diverge by accident. In the case of `Buffer::set_span` and `impl Widget for Span` even actively diverge. Which can create subtle bugs as people fine one and use it. (Argument with Rust: Rust decided against the ternary operator as the `if` can do the same.)

Multiple ways of doing the same creates complexity in maintaining the code and in reading code of other people. It can create subtle bugs. Both ways need to be tested, both ways need to be remembered when adapting things.

Boilerplate:

```diff
-Terminal::new(backend);
+backend.into_terminal();
```

This isn't shorter, just using the same method differently. Abstracting a lot of crossterm specific methods as `CrosstermBackend` methods is great. When we need to abstract our own methods away (`Terminal::new`) then there is something fishy. Is `Terminal` needed in the first place then? Could it be abstracted away completely?

Also: you like to search GitHub for code usage. Having exactly a single way of how to do things is way easier to search for.


> * you've previously mentioned introducing code that helps write more concise examples. However I don't think there should be example specific code in Ratatui. I.e. the code used in examples should be the same used in general user's apps
> 
> * this approach simplifies the necessary boilerplate by a significant amount

I think you are mixing up many methods which abstract crossterm related methods which are great and simplify using the `CrosstermBackend` by a lot and the `into_terminal` method which creates a duplicated way of creating a Terminal with the given Backend. I like the simplified usage of crossterm related methods by putting them as with_* methods. The into_terminal method is exactly the same as before, just different but introducing two ways of doing it.


Personally I would like to see either `into_terminal` or `Terminal::new` but not both. Maybe `Terminal::new` is more annoying to use than `into_terminal`, but then it should be marked as deprecated and the new way embraced.
One way should be used in examples, then its also discoverable how to use it.
If it feels like boilerplate, then it should be removed transparently for the user and only be some internal logic.

I agree with this PR in removing a lot of code from examples and real world code and I really look forward to using it, but having `into_terminal` and `Terminal::new` at the same time seem to be a step too much.

Reply:
I'll remove this method to get this PR over the line.

Reply:
9f608f954